File: src/main/java/com/key/oa/service/impl/UserDetailsServiceImpl.java
package com.key.oa.service.impl;

import com.key.oa.domain.LoginEmployee;
import com.key.oa.entity.Resource;
import com.key.oa.entity.Employee;
import com.key.oa.entity.Role;
import com.key.oa.service.EmployeeService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.*;

/**
 * @author 孙强
 * 实现UserDetailsService，从数据库中查找数据
 */
@Slf4j
@Service
public class UserDetailsServiceImpl implements UserDetailsService {
    private final EmployeeService employeeService;

    @Autowired
    public UserDetailsServiceImpl(EmployeeService employeeService) {
        this.employeeService = employeeService;
    }

    @Override
    public UserDetails loadUserByUsername(String jobNumber) throws UsernameNotFoundException {
        // 根据登录的工号查询员工信息
        Employee employee = employeeService.findByJobNumber(jobNumber);
        if (employee == null) {
            throw new UsernameNotFoundException(String.format("No employee found with job number: %s.", jobNumber));
        }
        Set<Role> reachableRoles = getReachableRoles(employee.getRoles());

        // 把员工信息封装成UserDetails对象返回
        return new LoginEmployee(employee, getAuthorities(reachableRoles));
    }

    private Set<Role> getReachableRoles(Collection<Role> roles) {
        if (roles == null || roles.isEmpty()) {
            return Collections.emptySet();
        }
        Set<Role> unprocessedRoles = new HashSet<>(10);
        unprocessedRoles.addAll(roles);

        Set<Role> reachableRoles = new HashSet<>(20);

        while (!unprocessedRoles.isEmpty()) {
            for (Role role : unprocessedRoles) {
                reachableRoles.add(role);

                //noinspection AlibabaDontModifyInForeachCircle
                unprocessedRoles.remove(role);
                unprocessedRoles.addAll(role.getChildren());
            }
        }

        return reachableRoles;
    }

    private List<String> getAuthorities(Set<Role> roles) {
        if (roles == null || roles.isEmpty()) {
            return Collections.emptyList();
        }

        List<Resource> authorities = new ArrayList<>();
        roles.forEach(role -> authorities.addAll(role.getResources()));
        return authorities.stream().map(Resource::getValue).toList();
    }
}

File: src/main/java/com/key/oa/service/impl/UserDetailsServiceImpl.kt
package com.key.oa.service.impl

import com.key.oa.dto.LoginEmployee
import com.key.oa.entity.Permission
import com.key.oa.entity.Role
import com.key.oa.service.EmployeeService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

@Service
class UserDetailsServiceImpl @Autowired constructor(private val employeeService: EmployeeService) : UserDetailsService {
    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(jobNumber: String): UserDetails {
        // 根据登录的工号查询员工信息
        val employee = employeeService.findByJobNumber(jobNumber)
            ?: throw UsernameNotFoundException(String.format("No employee found with job number: %s.", jobNumber))
        val reachableRoles = getReachableRoles(employee.roles)

        // 把员工信息封装成UserDetails对象返回
        return LoginEmployee(employee, getAuthorities(reachableRoles))
    }

    private fun getReachableRoles(roles: Collection<Role>): Set<Role> {
        if (roles.isEmpty()) {
            return emptySet()
        }

        val unprocessedRoles: MutableSet<Role> = HashSet(10)
        unprocessedRoles.addAll(roles)
        val reachableRoles: MutableSet<Role> = HashSet(20)
        while (unprocessedRoles.isNotEmpty()) {
            for (role in unprocessedRoles) {
                reachableRoles.add(role)
                unprocessedRoles.remove(role)
                unprocessedRoles.addAll(role.children)
            }
        }
        return reachableRoles
    }

    private fun getAuthorities(roles: Set<Role>): List<String> {
        if (roles.isEmpty()) {
            return emptyList()
        }

        val authorities: MutableList<Permission> = mutableListOf()
        roles.forEach { authorities.addAll(it.permissions) }
        return authorities.map { it.value }.toList()
    }
}